src/test_utils.rs

```diff
@@ -746,6 +747,7 @@ pub struct SentMessage {
     pub payload: String,
     recipients: String,
     pub sender_msg_id: MsgId,
+    sender_context: Context,
```

Comment:
isn't this a huge departure from the deltachat model?  afaik this isn't done anywhere currently.  not saying this is a bad idea, I always kind of wondered why we didn't go for putting `&Context` in there but it was never attempted.  cloning is... creative.

Reply:
> cloning is... creative.

It's an Arc, and we are cloning Contexts everywhere.

> afaik this isn't done anywhere currently.

That's true, like, we nowhere put Contexts into stucts afaik

Reply:
> > cloning is... creative.
> 
> No, we are cloning Contexts everywhere. It's an Arc, so it's fine.

Yes, I know it's the inner is in an Arc.  still, Arc cycles do scare me.

> > afaik this isn't done anywhere currently.
> 
> That's true, like, we nowhere put Contexts into stucts afaik

So I'm kind off -1 on starting to do this, **especially** with clones.  The current behaviour of passing it in explicitly results in very few clones of the Context because almost everywhere we use `&Context`, including in the calls replaced here.  Making this an owned reference now without any good motivation (I mean, it doesn't solve an awkward problem that's otherwise difficult to solve) does result in less efficient code and will make it much harder in the future to do this properly with storing the reference to the context in structs.  Right now having everything as references keeps the code better structured and avoids building in unexpected cycles.

Reply:
> Arc cycles do scare me.

That's a file that's only compiled in tests, so you would need to try quite hard to store it in a Context to build an Arc cycle.

> Making this an owned reference now without any good motivation [...] does result in less efficient code

These are tests, incrementing and decrementing an Arc counter a few times doesn't matter. Also, most of the time is spent in database accesses, not low-level stuff like Arc counters.

And, in fact, something equivalent to Arc-cloning is what you would do in almost all other programming languages. I love Rust for being able to have the extra security _and_ speed of not doing it, but think that it's overkill for tests. 

> I mean, it doesn't solve an awkward problem that's otherwise difficult to solve

The reason that I'm pushing back on this a bit is that lifetimes do make the code more difficult to read for newcomers, and sometimes

Reply:
> Arc cycles do scare me.

That's a file that's only compiled in tests, so you would need to try quite hard to store it in a Context to build an Arc cycle.

> Making this an owned reference now without any good motivation [...] does result in less efficient code

These are tests, incrementing and decrementing an Arc counter a few times doesn't matter. Also, most of the time is spent in database accesses, not low-level stuff like Arc counters.

And, in fact, something equivalent to Arc-cloning is what you would do in almost all other programming languages. I love Rust for being able to have the extra security _and_ speed of not doing it, but think that it's overkill for tests. 

> I mean, it doesn't solve an awkward problem that's otherwise difficult to solve

The reason that I'm pushing back on this a bit is that lifetimes do make the code more difficult to read for newcomers, and am trying to use the advanced Rust features as little as possible (not that lifetimes were the most advanced Rust feature we're using, so it will only help about this particular piece of code, and if you insist, we can also use lifetimes instead of clones here).

Reply:
I do think that we should have a `Context` or `&Context` in the struct at all, to prevent the caller from passing in the wrong context, e.g.:
```rust
        let sent_msg = alice.pop_sent_msg().await;
        let message = sent_msg.load_from_db(&bob).await;
```

This will in most cases even return a message (and not even panic), just the wrong one.

Reply:
Whoops, sorry for the tardy responses.

I agree the mixing of contexts is a concern, and this is more likely to happen in tests (currently at least) then elsewhere.

(As an aside, I also missed this was test-only code and I doubt I'll be the only one that misses this.  I still think starting to do this in one place will lead to doing this in many places, also outside of tests and in that case I'd rather get the pattern right.  This is also why I care an outside amount about something so trivial, I think it will set patterns.)

However, I still don't like starting to clone Contexts in structs.  Not because of the overheads of reference counting or anything.  But because of the static validation the compiler gives us on having the lifetimes.  Lifetimes are not just there to make the code a bit faster, they're there to help us write fewer bugs by not unexpectedly keeping things alive.  If we start cloning the context in structs relying on the context those structs keep the context alive.  We know this is not needed because right now all functions take the context by reference.  Furthermore we know it is possible to close a context or stop its I/O threads and things like that.  All of that results in complexity, and thus more bugs, creeping into the code.  So really, lifetimes help us write more correct code and we should use them and not avoid it.

And going back to the concern of mixing contexts: totally a good thing to try and fix.  I think the right approach is to use a struct with a lifetime that takes `context: &Context` as a field though.  Not clones.

Reply:
I'm going to go with lifetimes then

Reply:
> > No, we are cloning Contexts everywhere. It's an Arc, so it's fine.
> 
> We usually do this when we move Context into async task. Ideally this should not be needed, because we know the task started from the context can't outlive the context.

Huh, fair.  Sadly all tokio task spawning API's require the task to be `'static` AFAIK.  Theoretically this is leading towards structured concurrency but practically I don't know how to improve on this.